### ISSUE TITLE
Entity:IsConstrained only works with constraints added with the constraint library, and doesn't update clientside unless called serverside

### ISSUE THREAD

**Issue by Kefta** [Facepunch/garrysmod-issues#3837]
``IsConstrained`` has a lot of flaws in implementation:  everything is stored in an ``Entity.Constrained`` table that is only updated when constraints are added through the constraint library, meaning the function won't respond to map constraints. It also only updates the "IsConstrained" NWBool whenever the function is called serverside, meaning the client can have extremely outdated or incorrect information about the entity's constraint state.

The best solution would be to move ``Entity:IsConstrained`` to the engine and update a networked variable whenever ``Entity:SetPhysConstraintObjects`` / ``CPhysConstraint::CreateConstraint`` is called for each constraint class. Then, the "ConstraintBroken" gameevent can be used to update the entity's constrained state by checking with ``IPhysicsObject::IsAttachedToConstraint``.

tl;dr pseudocode:
```
void SetEntityConstraint( const hl_constraint_info_t &info )
{
	CBaseEntity *pEntity = dynamic_cast<CBaseEntity*>( info.pObjects[0]->GetGameData() );
	
	if ( pEntity != NULL )
		SetNetworkedBool( pEntity, "IsConstrained", true );
}

// Subclasses will have to be modified to call the constraint method
// (example for CPhysHinge)
IPhysicsConstraint *CreateConstraint( IPhysicsConstraintGroup *pGroup, const hl_constraint_info_t &info )
{	
	//...

	IPhysicsConstraint *pConstraint = physenv->CreateHingeConstraint( info.pObjects[0], info.pObjects[1], pGroup, m_hinge );
	
	if ( pConstraint != NULL )
		SetEntityConstraint( info );
	
	return pConstraint;
}

// Add to CBaseEntity's datadesc
DEFINE_INPUTFUNC( FIELD_VOID, "ConstraintBroken", InputOnBreak )

// Add to CBaseEntity
void InputOnBreak( inputdata_t &inputdata )
{
	IPhysicsObject *pPhysObj = VPhysicsGetObject();
	
	if ( pPhysObj != NULL && !pPhysObj->IsAttachedToConstraint( false ) )
		SetNetworkedBool( pEntity, "IsConstrained", false );
}
```

**Comment by Kefta** [comment]
I'm unsure if InputOnBreak is called upon physobj destruction, so you might have to check for that specifically.
